### ISSUE TITLE
C++ OpenTelemetry Plugin

### ISSUE THREAD

**Issue by wjones127** [grpc/grpc#32508]
### Is your feature request related to a problem? Please describe.

There is an OpenCensus plugin that collects metrics and spans with internal gRPC APIs. Would the project be open to having a parallel OpenTelemetry plugin?

### Describe the solution you'd like

An OpenTelemetry plugin similar to the OpenCensus one.

### Describe alternatives you've considered

Building a plugin externally doesn't seem possible, given the required APIs are private.

### Additional context

We maintain a framework that wraps gRPC C++, called Arrow Flight. Right now we provide OpenTelemetry support as a middleware in our layer ([based on this example](https://github.com/open-telemetry/opentelemetry-cpp/blob/main/examples/grpc/server.cc)), but it some cases users need more insight into the performance of gRPC itself, which is a total black box right now.

**Comment by yashykt** [comment]
Hi, we are preparing a roadmap for this. Please stay posted! :)

**Comment by lidavidm** [comment]
Is there any update on the roadmap?

**Comment by lidavidm** [comment]
It appears some of this was done in #33650. Are there plans for further work? Is the call tracer API meant for public use?

**Comment by ianmcook** [comment]
As noted at https://github.com/grpc/grpc-go/issues/6305, it was announced on 2023-05-01 that [OpenCensus is being sunsetted](https://opentelemetry.io/blog/2023/sunsetting-opencensus/) and also that OpenTelemetry now has feature parity with it for C++. We would love to have some more information about the timeline for OTel support in gRPC C++. Thanks

**Comment by lidavidm** [comment]
It looks like the current implementation records metrics, but not spans. And it seems OpenCensus is still artificially hardcoded into ServerContextBase, so other implementations can't start a span and share it with the RPC handler.

**Comment by lidavidm** [comment]
It's been another 6 months. @yashykt is there any update for when OpenTelemetry support might happen, or if gRPC exposes enough hooks to do this ourselves? (The interceptor mechanism is insufficient since I don't see a way to pass a span from the interceptor to the RPC handler.)

**Comment by jduo** [comment]
Alternatively would you be open to an external PR for this @yashykt ?

**Comment by yashykt** [comment]
Sorry for not posting an update here earlier!

The gRFC for OpenTelemetry metrics is available at https://github.com/grpc/proposal/blob/master/A66-otel-stats.md
This was implemented in gRPC C++ and has been available since 1.61.x

**Comment by yashykt** [comment]
The gRFC for OpenTelemetry tracing is in progress in https://github.com/grpc/proposal/pull/389. Please take a look. It'll be finished in the coming quarters.

There are some additional metrics also being designed in https://github.com/grpc/proposal/pull/421 and https://github.com/grpc/proposal/pull/419. These might be of interest as well.

**Comment by lidavidm** [comment]
Thanks for the context.

It seems like plugins/call filters are able to propagate state to RPC handlers and/or guarantee that they run in the same thread as the RPC handler? (Since the OpenTelemetry current span is a thread local.)

**Comment by yashykt** [comment]
> It seems like plugins/call filters are able to propagate state to RPC handlers and/or guarantee that they run in the same thread as the RPC handler? (Since the OpenTelemetry current span is a thread local.)

Is your query around how gRPC will propagate contexts from the application to be used as the parent context for the RPCs? As far as I see it, there are two ways we can support this - 
1) Explicitly provide a way to set the OpenTelemetry context. A similar thing was done for OpenCensus - https://github.com/grpc/grpc/blob/a19e01c4fd947ed5f4cc4a91412c4cbf029d2526/include/grpcpp/client_context.h#L373
2) Internally make sure that we propagate the parent OpenTelemetry context to the internal call tracers so that we preserve the parent context even when there is a thread hop.

**Comment by lidavidm** [comment]
Yes, or more specifically: how can we propagate state (like context) from an interceptor (or other mechanism that gets to run before/after every call) to the RPC handler itself?

I had seen that OpenCensus was privileged, if there is no way to do things like this besides modifying gRPC itself that is extremely unfortunate. (For instance, another thing we're trying to do is provide a session cookie, so we have to have some way of communicating the session from an interceptor to the RPC handler...hardcoding won't work since I doubt we'd be able to upstream that to gRPC.)

**Comment by yashykt** [comment]
> Yes, or more specifically: how can we propagate state (like context) from an interceptor (or other mechanism that gets to run before/after every call) to the RPC handler itself?

This has come up before in various contexts. We've thought of working on a context propagation mechanism that does not use metadata.
The C++ interception API is also going to go through an overhaul so maybe we can consider this @ctiller

**Comment by lidavidm** [comment]
That's great news! We've been struggling with a pile of hacks because of that and an interception API that is able to pass context to the RPC handler somehow (same as in Java and other gRPC implementations) would be much appreciated.

**Comment by yashykt** [comment]
The gRPC C++ OpenTelemetry plugin now supports metrics and tracing. What I still want to build is a general purpose per-call context mechanism that can be shared between the application and the interceptors, which I'm going to track in https://github.com/grpc/grpc/issues/36639
